### DISCUSSION TITLE
Help! Drop-outs on MacOS

### DISCUSSION THREAD

**Discussion by EdBeardon** [lathoub/Arduino-AppleMIDI-Library#137] [Q&A] [answered]
Hi Guys, 

Wondered if I could get some help with my code - I've used one of the examples and managed to get it to send a MIDI  note continuously 

My Mac is able to connect to the session and receive these MIDI notes but after 2/3 minutes my connection to the RTP session times out. Wondered if anyone had experienced this and could help?

I'm using a W5500 and the Ethernet3 library. 

Code & image attached 

![Screenshot 2021-10-31 at 10 15 11](https://user-images.githubusercontent.com/93470919/139577904-02f39e68-ca3e-4f54-9f43-d3cc337fe338.png)

#include <Ethernet3.h>
#define USE_DIRECTORY
#define SerialMon Serial
#define APPLEMIDI_DEBUG SerialMon
#include <AppleMIDI.h>

// Enter a MAC address for your controller below.
// Newer Ethernet shields have a MAC address printed on a sticker on the shield
byte mac[] = {
  0x82, 0x54, 0xAD, 0xF6, 0xB2, 0x4D
};

unsigned long t1 = millis();
int8_t isConnected = 0;

APPLEMIDI_CREATE_DEFAULTSESSION_INSTANCE();

void(* resetFunc) (void) = 0;  // declare reset fuction at address 0

// -----------------------------------------------------------------------------
//
// -----------------------------------------------------------------------------
void setup()
{

  DBG_SETUP(115200);
  DBG("Booting");

  Ethernet.init(4); // maxSockNum = 4 Socket 0...3 -> RX/TX Buffer 4k

  if (Ethernet.begin(mac) == 0) {
    DBG(F("Failed DHCP, check network cable & reboot"));
    for (;;);
  }

  AppleMIDI.directory.push_back(IPAddress(192, 168, 0, 177));
  AppleMIDI.directory.push_back(IPAddress(192, 168, 0, 66));
//  AppleMIDI.whoCanConnectToMe = APPLEMIDI_NAMESPACE::None;
//  AppleMIDI.whoCanConnectToMe = APPLEMIDI_NAMESPACE::OnlyComputersInMyDirectory;
  AppleMIDI.whoCanConnectToMe = APPLEMIDI_NAMESPACE::Anyone;

  DBG(F("OK, now make sure you an rtpMIDI session that is Enabled"));
  DBG(F("Add device named Arduino with Host"), Ethernet.localIP(), "Port", AppleMIDI.getPort(), "(Name", AppleMIDI.getName(), ")");
  DBG(F("Select and then press the Connect button"));
  DBG(F("Then open a MIDI listener and monitor incoming notes"));

  MIDI.begin();

  // Stay informed on connection status
  AppleMIDI.setHandleConnected([](const APPLEMIDI_NAMESPACE::ssrc_t & ssrc, const char* name) {
    isConnected++;
    DBG(F("Connected to session"), ssrc, name);
  });
  AppleMIDI.setHandleDisconnected([](const APPLEMIDI_NAMESPACE::ssrc_t & ssrc) {
    isConnected--;
    DBG(F("Disconnected"), ssrc);
    resetFunc(); //call reset
  });

  MIDI.setHandleNoteOn([](byte channel, byte note, byte velocity) {
    DBG(F("NoteOn"), note);
  });
  MIDI.setHandleNoteOff([](byte channel, byte note, byte velocity) {
    DBG(F("NoteOff"), note);
  });

  DBG(F("Sending MIDI messages every second"));
}

// -----------------------------------------------------------------------------
//
// -----------------------------------------------------------------------------
void loop()
{
  // Listen to incoming notes
  MIDI.read();

  MIDI.sendNoteOn(01,55,01);
    //    MIDI.sendNoteOff(note, velocity, channel);
  //}

  delay(100);
}

**Comment by lathoub** [comment, marked as answer]
Hi @EdBeardon 

there are many issues with the use of `delay(100);` in the main loop. Don;t do this! Look at [this example](https://github.com/lathoub/Arduino-AppleMIDI-Library/blob/master/examples/AVR_NoteOnOffEverySec/AVR_NoteOnOffEverySec.ino) to send notes every x seconds

Start with removing `delay(100);` and `MIDI.sendNoteOne` and see if it remains connected

**Reply by EdBeardon** [reply]
Thanks @lathoub, still seems to be connected without sending any messages - what's the next step please?

**Reply by lathoub** [reply]
Good! 
Next: look at [this example](https://github.com/lathoub/Arduino-AppleMIDI-Library/blob/master/examples/AVR_NoteOnOffEverySec/AVR_NoteOnOffEverySec.ino) on how to send notes every x seconds

**Reply by EdBeardon** [reply]
Thanks! Got it working, without disconnecting with code below:

if ((isConnected > 0) && (millis() - t1) > 1000)
  {
    t1 = millis();
    
    MIDI.sendNoteOn(1,55,1);
//    MIDI.sendNoteOff(note, velocity, channel);

**Reply by lathoub** [reply]
Let me know if `Ethernet3.h` is stable
